src/bindings/python/src/pyopenvino/graph/model.cpp

```diff
@@ -260,52 +261,68 @@ void regclass_graph_Model(py::module m) {
 
     function.def(
         "reshape",
-        [](ov::Model& self, const ov::PartialShape& partial_shape) {
-            self.reshape(partial_shape);
+        [](ov::Model& self, const py::handle& partial_shapes) {
+            if (py::isinstance<ov::PartialShape>(partial_shapes)) {
+                self.reshape(partial_shapes.cast<ov::PartialShape>());
+            } else if (py::isinstance<py::dict>(partial_shapes)) {
+                std::map<ov::Output<ov::Node>, ov::PartialShape> new_shapes;
+                for (const auto item : partial_shapes.cast<py::dict>()) {
+                    std::pair<ov::Output<ov::Node>, ov::PartialShape> new_shape;
+                    // check keys
+                    if (py::isinstance<py::int_>(item.first)) {
+                        new_shape.first = self.input(item.first.cast<size_t>());
+                    } else if (py::isinstance<py::str>(item.first)) {
+                        new_shape.first = self.input(item.first.cast<std::string>());
+                    } else if (py::isinstance<ov::Output<ov::Node>>(item.first)) {
+                        new_shape.first = item.first.cast<ov::Output<ov::Node>>();
+                    } else {
+                        throw py::type_error("Incorrect key " + std::string(item.first.get_type().str()) +
+                                             " type to reshape a model.");
+                    }
+                    // check values
+                    if (py::isinstance<ov::PartialShape>(item.second)) {
+                        new_shape.second = item.second.cast<ov::PartialShape>();
```

Comment:
Some time ago, @popovaan implemented a single way to serialize/deserialize PartialShape to/from string. Can we reuse this stuff here and also accept partial shape as a string. It will enable a common language between MO/diagnostics output that consists of shapes, for example we can use '[?, 3, 224, 224]' or '[1, 1..128]' to just set a dynamic batch exactly in the same form that MO should accept without forcing user to learn a new syntax for ranges. @popovaan we did that thing for PartialShape, right? Can it be reused here?

Reply:
It seems to be a good idea if the reshape method approves dynamic batches.

Reply:
Yes, str_to_partial_shape() method can be used here:
https://github.com/openvinotoolkit/openvino/blob/541627d319dc5e12262835e4f830f02095dfc8a1/src/frontends/ir/src/utils.cpp#L69

Reply:
Note that str_to_partial_shape() expects shape without brackets, for example '?, 3, 224, 224'

Reply:
To do not add ov_ir_frontend dependency to python bindings we will implement str_to_partial_shape on our side.

Reply:
done